File: langs/csharp/src/VMx/Hierarchical/TreeStructureChangedMessage.cs
using VMx.Components;
using VMx.Messages;

namespace VMx.Hierarchical;

/// <summary>
/// Discriminated enum for the structural mutation that occurred in a
/// <see cref="HierarchicalVM{TModel,TVM}"/> subtree.
/// </summary>
public enum TreeStructureChange
{
    /// <summary>A child node was added to the subtree.</summary>
    Added,

    /// <summary>A child node was removed from the subtree.</summary>
    Removed,

    /// <summary>A node was moved to a different parent within the tree.</summary>
    Reparented,
}

/// <summary>
/// Message published on <c>IMessageHub</c> when a <see cref="HierarchicalVM{TModel,TVM}"/>
/// subtree changes structurally (add / remove / reparent of a child).
///
/// See spec/18-hierarchical-vm.md §6 and ADR-0028 §3.4.
/// </summary>
/// <param name="Source">The node whose <c>Children</c> collection changed.</param>
/// <param name="Change">The kind of structural mutation.</param>
/// <param name="Affected">The node that was added, removed, or reparented.</param>
/// <param name="Index">Position in <c>Children</c> at which the change occurred; -1 when not applicable (e.g. Reparented).</param>
public sealed record TreeStructureChangedMessage(
    object Source,
    TreeStructureChange Change,
    object Affected,
    int Index) : IMessage
{
    /// <inheritdoc/>
    /// <remarks>
    /// Reads <see cref="IComponentVM.Name"/> when the source implements
    /// <see cref="IComponentVM"/> (the canonical case), matching Python
    /// (`hierarchical_vm.py`) and TypeScript (`hierarchicalVm.ts`) which both
    /// emit the VM's configured name. Falls back to the runtime type name
    /// for the rare case where a non-VM source is supplied.
    /// </remarks>
    public string SenderName => (Source as IComponentVM)?.Name ?? Source.GetType().Name;

    /// <summary>The source cast to <see cref="object"/> (satisfies <see cref="IMessage"/> protocol).</summary>
    public object SenderObject => Source;
}

File: langs/swift/Sources/VMx/Messages/TreeStructureChangedMessage.swift
//
// TreeStructureChangedMessage — published on the message hub when a
// HierarchicalVM subtree changes structurally (add / remove / reparent).
//
// See spec/18-hierarchical-vm.md §6 and ADR-0028 §3.4.
//
import Foundation

/// Discriminated enum for the structural mutation that occurred.
public enum TreeStructureChange: Sendable, Equatable {
    case added
    case removed
    case reparented
}

/// Message published when a HierarchicalVM's children list mutates.
///
/// `sender` (via `senderObject`) is the node whose children list changed
/// (the parent that called `addChild` / `removeChild` / `reparentChild`).
/// `affected` is the child node that was added, removed, or reparented.
/// `index` is the position in the children list; -1 when not applicable
/// (reparent).
public struct TreeStructureChangedMessage: Message {
    public let senderObject: AnyObject
    public let senderName: String
    public let change: TreeStructureChange
    public let affected: ComponentVMBase
    /// Index in the children list at which the change occurred; -1 for reparent.
    public let index: Int

    public init(
        sender: AnyObject,
        senderName: String,
        change: TreeStructureChange,
        affected: ComponentVMBase,
        index: Int
    ) {
        self.senderObject = sender
        self.senderName = senderName
        self.change = change
        self.affected = affected
        self.index = index
    }

    /// Spec alias of `senderObject`. Matches the `PropertyChangedMessage`
    /// convention.
    public var sender: AnyObject { senderObject }
}
